src/main/java/org/graylog/integrations/pagerduty/client/MessageFactory.java

```diff
@@ -0,0 +1,128 @@
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Factory class for PagerDuty messages, heavily based on the works of the cited authors.
+ *
+ * @author Jochen Schalanda
+ * @author James Carr
+ * @author Dennis Oelkers
+ * @author Padma Liyanage
+ * @author Edgar Molina
+ */
+public class MessageFactory {
+    private static final List<String> PAGER_DUTY_PRIORITIES = Arrays.asList("info", "warning", "critical");
+
+    private final StreamService streamService;
+    private final EventNotificationService eventNotificationService;
+
+    @Inject
+    MessageFactory(StreamService streamService, EventNotificationService eventNotificationService) {
+        this.streamService = streamService;
+        this.eventNotificationService = eventNotificationService;
+    }
+
+    public PagerDutyMessage createTriggerMessage(EventNotificationContext ctx) {
+        final ImmutableList<MessageSummary> backlog = eventNotificationService.getBacklogForEvent(ctx);
+        final EventNotificationModelData modelData = EventNotificationModelData.of(ctx, backlog);
+        final PagerDutyNotificationConfig config = (PagerDutyNotificationConfig) ctx.notificationConfig();
+
+        String eventTitle = modelData.eventDefinitionTitle();
+        String eventPriority = PAGER_DUTY_PRIORITIES.get(0);
+        int priority = ctx.eventDefinition().get().priority() - 1;
+        if (priority >= 0 && priority <= 2) {
+            eventPriority = PAGER_DUTY_PRIORITIES.get(priority);
+        }
+
+        List<Link> streamLinks =
+                streamService
+                        .loadByIds(modelData.event().sourceStreams())
+                        .stream()
+                        .map(stream -> buildStreamWithUrl(stream, ctx, config))
+                        .collect(Collectors.toList());
+
+        String dedupKey = "";
+        if (config.customIncident()) {
+            dedupKey = String.format(
+                    "%s/%s/%s", config.keyPrefix(), modelData.event().sourceStreams(), eventTitle);
+        }
+
+
+        Map<String, String> payload = new HashMap<String, String>();
+        payload.put("summary", modelData.event().message());
+        payload.put("source", "Graylog:" + modelData.event().sourceStreams());
+        payload.put("severity", eventPriority);
+        payload.put("timestamp", modelData.event().eventTimestamp().toString());
+        payload.put("component", "GraylogAlerts");
+        payload.put("group", modelData.event().sourceStreams().toString());
+        payload.put("class", "alerts");
+
+        return new PagerDutyMessage(
+                config.routingKey(),
+                "trigger",
+                dedupKey,
+                config.clientName(),
+                config.clientUrl(),
+                streamLinks,
+                payload);
+    }
+
+    private Link buildStreamWithUrl(Stream stream, EventNotificationContext ctx, PagerDutyNotificationConfig config) {
```

Comment:
do we need this method ?

Reply:
It is being used.  I kept it because the intent here is to bring the PagerDuty Notification from the Graylog Labs repo into our official release without changing the functionality.